src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -795,203 +789,187 @@ where
-                CapabilitySet::new()
-            };
-            // Explicitly release the unneeded capabilities!
-            capabilities.clear();
-
-            let upper_ts = resume_upper.as_option().copied().unwrap();
-
-            // NOTE: Below, we allow compaction only up to around 3 days before
-            // the resumption frontier. We still restart from right before the
-            // upper which is a) perfectly safe to do, and b) will make extra
-            // certain that the `as_of` is valid with regards to the since. If
-            // we pick the same time that we compact up to, we might run into
-            // the same (alleged/potential) off-by-one bug that this
-            // three-day-bandaid below is trying to paper over.
-            let as_of = Antichain::from_elem(upper_ts.saturating_sub(1));
-            let mut timestamper = match ReclockOperator::new(
-                Arc::clone(&persist_clients),
-                storage_metadata.clone(),
-                now.clone(),
-                timestamp_interval.clone(),
-                as_of,
-            )
-            .await
-            {
-                Ok(t) => t,
-                Err(e) => {
-                    panic!("Failed to create source {} timestamper: {:#}", name, e);
-                }
-            };
-            // The global view of the source_upper, which we track by combining
-            // summaries from the raw reader operators.
-            let mut global_source_upper = timestamper
-                .source_upper_at_frontier(resume_upper.borrow())
-                .expect("source_upper_at_frontier to be used correctly");
-
-            if active_worker {
-                let new_ts_upper = timestamper
-                    .reclock_frontier(&global_source_upper)
-                    .expect("compacted past upper");
-
-                cap_set.downgrade(new_ts_upper);
-
-                // Emit initial snapshot of the remap_shard, bootstrapping
-                // downstream reclock operators.
-                let remap_trace = timestamper.remap_trace();
-                trace!(
-                    "remap({id}) {worker_id}/{worker_count}: \
-                    emitting initial remap_trace. \
-                    source_upper: {:?} \
-                    trace_updates: {:?}",
-                    global_source_upper,
-                    remap_trace
-                );
+    let (shutdown_tx, shutdown_rx) = tokio::sync::oneshot::channel::<()>();
 
-                let mut remap_output = remap_output.activate();
-                let cap = cap_set.delayed(cap_set.first().unwrap());
-                let mut session = remap_output.session(&cap);
-                session.give(remap_trace);
+    remap_op.build(move |mut capabilities| async move {
+        if !active_worker {
+            return;
+        }
+
+        let mut buffer = Vec::new();
+        let mut cap_set = CapabilitySet::from_elem(capabilities.pop().expect("missing capability"));
```

Comment:
nit: should this use `itertools::exactly_one`?

Reply:
done